build: create a Make target for `go generate`

benesch opened the pull request:
PR #14629 adjusted our Makefile to install tools into REPO_ROOT/bin
instead of GOBIN. This change appears to have broken builds on fresh
TeamCity agents. Because `go generate` was previously invoked outside
the Makefile, `go generate` ran with a PATH that did not include
REPO_ROOT/bin.

This commit adds a `generate` target to the Makefile that ensures `go
generate` is run with the correct PATH. It has the added benefit of
symmetry with the other Make targets that shadow Go tools, like `make
build`, `make install`, and `make test`.

This commit additionally adjusts `make check` and `make generate` to
depend on `gotestdashi`; there was previously an implicit dependency
here that could lead to hard-to-track-down failures. (See the comment
that was moved from teamcity-check.sh to the Makefile for details.)

benesch committed:
build: create a Make target for `go generate`

PR #14629 adjusted our Makefile to install tools into REPO_ROOT/bin
instead of GOBIN. This change appears to have broken builds on fresh
TeamCity agents. Because `go generate` was previously invoked outside
the Makefile, `go generate` ran with a PATH that did not include
REPO_ROOT/bin.

This commit adds a `generate` target to the Makefile that ensures `go
generate` is run with the correct PATH. It has the added benefit of
symmetry with the other Make targets that shadow Go tools, like `make
build`, `make install`, and `make test`.

This commit additionally adjusts `make check` and `make generate` to
depend on `gotestdashi`; there was previously an implicit dependency
here that could lead to hard-to-track-down failures. (See the comment
that was moved from teamcity-check.sh to the Makefile for details.)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14672)
<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Makefile, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2xUqw4PlazkvjKDny:-Kh2xUqw4PlazkvjKDnz:b6hcs5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L243)):*
> ```Makefile
> # x/tools/go/loader is fragile (this includes stringer, vet and others).
> #
> # The blocking issue is https://github.com/golang/go/issues/14120; see
> ```

consider breaking this comment into 3 parts - the first part which applies to both check and generate, and then separate ones for the two, listing the issues.

perhaps put generate above check for better grouping.

---

*[Makefile, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2xhemDdf_onpVdr2e:-Kh2xhemDdf_onpVdr2f:b-tz0y4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L254)):*
> ```Makefile
> checkshort: override TAGS += check
> checkshort: gotestdashi
> 	$(GO) test ./build -v -tags '$(TAGS)' -short -run 'TestStyle/$(TESTS)'
> ```

ugh, can you add `ldflags` to this? i missed it in an earlier change.

---

*[pkg/security/securitytest/test_certs/README.md, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2y4JyxvoQ1B4jKD35:-Kh2y4JyxvoQ1B4jKD36:b19qzwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/pkg/security/securitytest/test_certs/README.md#L27)):*
> ```Markdown
> ./cockroach cert --ca-cert=pkg/security/securitytest/test_certs/ca.crt --ca-key=pkg/security/securitytest/test_certs/ca.key --cert=pkg/security/securitytest/test_certs/client.root.crt --key=pkg/security/securitytest/test_certs/client.root.key create-client root
> ./cockroach cert --ca-cert=pkg/security/securitytest/test_certs/ca.crt --ca-key=pkg/security/securitytest/test_certs/ca.key --cert=pkg/security/securitytest/test_certs/client.testuser.crt --key=pkg/security/securitytest/test_certs/client.testuser.key create-client testuser
> make generate ./pkg/security/securitytest
> ```

doesn't this need to be `make generate PKG=...`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-:-Kh2yDa1xbTn-_prvkdQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented on pkg/ui/Makefile:
```diff
@@ -27,8 +27,8 @@ all: lint test $(GOBINDATA_TARGET)
 
-# Running `go generate` will call this target. Update this if you add new
-# generated files.
+# Running `make generate` from the root will call this target via `go generate`.
```
The UI tests are no longer run via `make generate`, so this should be deleted.

tamird replied:
What do the UI tests have to do with this?

jordanlewis replied:
This diff is in `pkg/ui/Makefile`, and it claims that running `make generate` from the root will trigger this target, which is inaccurate.

tamird replied:
How is that inaccurate? `make generate` from the root will run `go generate`, which will exactly run this target https://github.com/cockroachdb/cockroach/blob/master/pkg/ui/ui.go#L21

jordanlewis replied:
You're right, I was mistakenly conflating the UI tests and the UI generate target (as you rightly pointed out above). Nevermind.

jordanlewis commented on CONTRIBUTING.md:
```diff
@@ -108,7 +108,7 @@ executable will be in your current directory and can be run as shown in the
 + Run the test suite locally:
 
-  `go generate ./pkg/... && make check test testrace`
+  `make generate && make check test testrace`
```
You didn't introduce this inaccuracy, but this is no longer reflective of the full test suite. The UI tests are invoked separately - currently I think the best way is `make -C pkg/ui` but perhaps we should add a new top-level target that does this for you.

jordanlewis replied:
Happy to see this left undone in this PR though.

tamird replied:
Also not sure this needs to discuss the UI tests - that probably belongs in that package's README.

jordanlewis replied:
I disagree - this stanza is explicitly described as "run the test suite locally". That implies that all tests will be run. Without running the UI tests, not all of the tests will be run.

tamird replied:
🤷

jordanlewis reviewed (commented):
Sweet! I have a few comments

benesch committed:
build: pass LINKFLAGS through to checkshort, too

8052b6df standardized our handling of LINKFLAGS across Go tool
invocations, but missed this one in `checkshort`.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2roDYTYuuy7OYwxQE-r1-111:-Kh32YCG4EKIfOtRo9Ql:b-77l48h) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/CONTRIBUTING.md#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

🤷 
</blockquote></details>

It's also no longer possible to run `test` and `testrace` in the same `make` invocation (but again, doesn't need to be fixed in this PR)

---

*[build/teamcity-check.sh, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh32sGn9sqEUiia3IOf:-Kh32sGn9sqEUiia3IOg:b-gyvxi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/build/teamcity-check.sh#L13)):*
> ```Shell
> build/builder.sh /bin/bash -c '! git status --porcelain | read || (git status; git diff -a 1>&2; exit 1)'
> 
> build/builder.sh make -C pkg/ui test
> ```

Comment that this logically belongs in `teamcity-test.sh` but we do it here because `teamcity-check.sh` is shorter than the regular tests and this minimizes total build time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-:-Kh33D6Ic80zMra5WekF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 2 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[CONTRIBUTING.md, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2roDYTYuuy7OYwxQE-r1-111:-Kh31pT-LWNDMCmwuVOt:ba1u7ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/CONTRIBUTING.md#L111)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's also no longer possible to run `test` and `testrace` in the same `make` invocation (but again, doesn't need to be fixed in this PR)
</blockquote></details>

Ok, punting on this for now. (I did clean the command up to avoid two separate Make invocations.)

---

*[Makefile, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2xUqw4PlazkvjKDny:-Kh32ljyZDZHeYxDrU9H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consider breaking this comment into 3 parts - the first part which applies to both check and generate, and then separate ones for the two, listing the issues.

perhaps put generate above check for better grouping.
</blockquote></details>

Done.

---

*[Makefile, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2xhemDdf_onpVdr2e:-Kh33YpHS74QnU10tKXJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ugh, can you add `ldflags` to this? i missed it in an earlier change.
</blockquote></details>

Done.

---

*[build/teamcity-check.sh, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh32sGn9sqEUiia3IOf:-Kh33Nu7CSsrCIPxEIBv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/build/teamcity-check.sh#L13)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that this logically belongs in `teamcity-test.sh` but we do it here because `teamcity-check.sh` is shorter than the regular tests and this minimizes total build time.
</blockquote></details>

Done.

---

*[pkg/security/securitytest/test_certs/README.md, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2y4JyxvoQ1B4jKD35:-Kh33QqEGZyXZKfJtVmK:b-k614gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/pkg/security/securitytest/test_certs/README.md#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

doesn't this need to be `make generate PKG=...`?
</blockquote></details>

Yes, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672)*
<!-- Sent from Reviewable.io -->

tamird commented:
Don't forget to remove `vols="${vols} --volume=${gocache}/bin/docker_amd64:/go/bin"` from builder.sh.

---

Reviewed 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[Makefile, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh36AYFiSKPgfb3bSss:-Kh36AYFiSKPgfb3bSst:buhl4i0) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L242)):*
> ```Makefile
> # and code generators reliably. More precisely, anything that using
> # x/tools/go/loader is fragile (this includes stringer, vet and others).
> #
> ```

keep this? my editor will overzealously re-wrap this paragraph otherwise.

---

*[pkg/security/securitytest/test_certs/README.md, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2y4JyxvoQ1B4jKD35:-Kh36O3vgNaunUtgD1-n:b-cv3a70) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/pkg/security/securitytest/test_certs/README.md#L27)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Yes, thanks.
</blockquote></details>

Looks unchanged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-:-Kh362qGkPSfTnvbf5RS:b-v9uo3i)*
<!-- Sent from Reviewable.io -->

tamird commented:
Actually, don't remove it.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-:-Kh36SX9CfEibDrSRUqM:bpchsvf)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[Makefile, line 242 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh36AYFiSKPgfb3bSss:-Kh3BidPdy8pSOeglZNq:b-3z3twl) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/Makefile#L242)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

keep this? my editor will overzealously re-wrap this paragraph otherwise.
</blockquote></details>

I think this is no longer relevant.

---

*[pkg/security/securitytest/test_certs/README.md, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672#-Kh2y4JyxvoQ1B4jKD35:-Kh3Bl_WXCBmbaONuW5E:b7nmjyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/399912f7243b426f5ccce8a461cf832fa46de9e8/pkg/security/securitytest/test_certs/README.md#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Looks unchanged.
</blockquote></details>

Actually done now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 5 files at r4, 5 of 5 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14672)*
<!-- Sent from Reviewable.io -->
